src/storage/aggregator.h

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2021 4Paradigm
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SRC_STORAGE_AGGREGATOR_H_
+#define SRC_STORAGE_AGGREGATOR_H_
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "codec/codec.h"
+#include "proto/tablet.pb.h"
+#include "proto/type.pb.h"
+#include "storage/table.h"
+
+namespace openmldb {
+namespace storage {
+
+using Dimensions = google::protobuf::RepeatedPtrField<::openmldb::api::Dimension>;
+using ::openmldb::type::DataType;
+enum class AggrType {
+    kSum = 1,
+    kMin = 2,
+    kMax = 3,
+    kCount = 4,
+    kAvg = 5,
+};
+
+enum class WindowType {
+    kRowsNum = 1,
+    kRowsRange = 2,
+};
+
+union AggrVal {
+    int16_t vsmallint;
+    int32_t vint;
+    int64_t vlong;
+    float vfloat;
+    double vdouble;
+};
+
+struct AggrBuffer {
+    AggrVal aggr_val_;
+    int64_t ts_begin_;
+    int64_t ts_end_;
+    int32_t aggr_cnt_;
+    uint64_t binlog_offset_;
+    AggrBuffer() : aggr_val_(), ts_begin_(-1), ts_end_(0), aggr_cnt_(0), binlog_offset_(0) {}
+    void clear() {
+        memset(&aggr_val_, 0, sizeof(aggr_val_));
+        ts_begin_ = -1;
+        ts_end_ = 0;
+        aggr_cnt_ = 0;
+        binlog_offset_ = 0;
+    }
+};
+struct AggrBufferLocked {
+    std::unique_ptr<std::mutex> mu_;
+    AggrBuffer buffer_;
```

Comment:
it's better to use consecutive buffer to save these field in order to reduce memory usage

Reply:
> it's better to use consecutive buffer to store these field in order to reduce memory usage

the union of `aggr_val` extra memory usage is small compared to the table space. and also need a pointer to local buffer space will increase memory usage.

Reply:
with [flexible array member](https://en.wikipedia.org/wiki/Flexible_array_member)

Reply:
#1578